npm/src/workers/common.ts

```diff
@@ -425,28 +448,57 @@ export function initService<
   TServiceEventMsg extends IServiceEventMessage,
 >(
   postMessage: (
-    msg: ResponseMessage<TService> | EventMessage<TServiceEventMsg>,
+    msg:
+      | ResponseMessage<TService>
+      | EventMessage<TServiceEventMsg>
+      | CommonEventMessage,
   ) => void,
   serviceProtocol: ServiceProtocol<TService, TServiceEventMsg>,
   wasm: Wasm,
   qscLogLevel?: number,
 ): (req: RequestMessage<TService>) => Promise<void> {
-  function telemetryHandler(telemetry: TelemetryEvent) {
+  function postTelemetryMessage(telemetry: TelemetryEvent) {
     postMessage({
-      messageType: "event",
+      messageType: "common-event",
       type: "telemetry-event",
       detail: telemetry,
     });
   }
 
+  function postLogMessage(level: number, target: string, ...args: any) {
+    let data = args;
+    try {
+      // Only structured cloneable objects can be sent in worker messages.
+      // Test if this is the case.
+      structuredClone(args);
+    } catch (e) {
+      // Uncloneable object.
+      // Use String(args) instead of ${args} to handle all possible values
+      // without throwing. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_coercion
+      data = ["unsupported log data " + String(args)];
+    }
+    postMessage({
+      messageType: "common-event",
+      type: "log",
+      detail: { level, target, data },
+    });
+  }
+
+  // Override the global logger
+  log.error = (...args) => postLogMessage(1, "worker", ...args);
+  log.warn = (...args) => postLogMessage(2, "worker", ...args);
+  log.info = (...args) => postLogMessage(3, "worker", ...args);
+  log.debug = (...args) => postLogMessage(4, "worker", ...args);
+  log.trace = (...args) => postLogMessage(5, "worker", ...args);
```

Comment:
Does this send all the messages from the worker without filtering on level first? That can be quite chatty.

Reply:
Ouch, yes, good catch. I'll fix this